### ISSUE TITLE
Shell: Dynamic Code

### ISSUE THREAD

**Issue by jaheba** [awslabs/gluon-ts#1766]
# GluonTS-shell Dynamic Code

## Summary

Add a mechanism to GluonTS-shell to run custom code, without the need to build and push a new container.

## Overview

Currently, if one wants to run GluonTS on SageMaker using `gluonts.shell`, a container needs to be build and published to ECR. Any code-change needs to go through this process before it's avaiable for use in SageMaker.

Thus, we want to add a mechanism that allows to dynamically use code-artefacts inside the container.

From a technical point of view, the solution is simple. Before execution, the desired artefacts are uploaded to s3. For training jobs they are then provided as part of the input data, and for inference code artefacts are added as part of the model by the training job.

## Client

When launching a training-job, users can reference files that should be uploaded.

These can be:

* Python modules (e.g. `my-file.py`)
* Python packages

Modules are just uploaded to the repository. Packages are built using `python setup.py sdist` to create source-distributions, which are then subsequently copied to s3.

Further, a manifest-file is generated (referencing the job's artefacts) and uploaded to the same location as the model-output on s3.

## S3 Repository

All artefacts are uploaded to a central repository: a user provided s3-prefix (e.g: `s3://my-experiments/code`). To enable caching, we prefix each artefact by its contents hash, so the structure of the repository looks like this:

```
s3://my-experiments/code/
    <hash 1>/object_1.py
    <hash 2>/object_2.py
    ...
```

We then generate a manifest for each training-job, which references objects in the repository:

```json
[
    {
        "prefix": "s3://my-experiments/code/"
    },
    "<hash>/object_1.py",
    "<hash>/object_2.tar.gz"
]
```

## Code Channel

Artefacts are added to training jobs, using a `code` channel. Using a manifest, SageMaker downloads the required artefact for us:

```
/opt/ml/input/data/code/
    <hash>/object_1.py
    <hash>/object_2.tar.gz
```

In inference, the code is located at:

```
/opt/ml/model/code/
    object_1.py
    object_2.tar.gz
```
    
We "install" each entry of the code-channel to `/opt/ml/code/`. Simple modules are just copied, whilst packages are properly installed using `pip install --target /opt/ml/code/ <package>`.

As a final step, add `/opt/ml/code` to `PYTHONPATH`  and restart the process using `os.execve`.

**Comment by mbohlkeschneider** [comment]
<3 that! Can we use the same code channel for training/inference. Or is there no data downloaded for inference? I'm not sure about that to be honest.

**Comment by mbohlkeschneider** [comment]
I also did not quite understand what the purpose of the hashing is?

**Comment by mbohlkeschneider** [comment]
Lastly, it would be nice to be able to track back what code has been used in a specific run. Would this still be in the manifest/would the code be somewhere referenced for download (as it is with the sagemaker estimator logic)?

**Comment by jaheba** [comment]
> I also did not quite understand what the purpose of the hashing is?

Two reasons, caching of code artefacts and a file-name independent prefix.

If I have a bigger dependency, the hash ensures that each version is only ever uploaded once. Also, if I upload different versions, I get different hashes as prefix.

**Comment by jaheba** [comment]
> Lastly, it would be nice to be able to track back what code has been used in a specific run. Would this still be in the manifest/would the code be somewhere referenced for download (as it is with the sagemaker estimator logic)?

Yes, this should be generally possible. However, there are two code sources: the container image and the code-channel. For the code-channel the manifest identifies the exact version, which are immutable. For the container you also need to make sure, that each change creates a new image tag to be really know what code has been executed.

**Comment by mbohlkeschneider** [comment]
> > I also did not quite understand what the purpose of the hashing is?
> 
> Two reasons, caching of code artefacts and a file-name independent prefix.
> 
> If I have a bigger dependency, the hash ensures that each version is only ever uploaded once. Also, if I upload different versions, I get different hashes as prefix.

So a new hash will be generated on the code (code change -> new hash?).

**Comment by jaheba** [comment]
Yes, we would take the hash of the file (just like git does).

**Comment by lostella** [comment]
Duplicate of #218?

**Comment by jaheba** [comment]
> Duplicate of #218?

Yes, although this is more of a RFC, while #218 can be understood as a tracking issue.
